cpp/include/ucxx/request_data.h

```diff
@@ -0,0 +1,208 @@
+  const ::ucxx::Tag _tag{0};          ///< Tag to match
+  const ::ucxx::TagMask _tagMask{0};  ///< Tag mask to use
+
+  /**
+   * @brief Constructor send tag-specific data.
+   *
+   * Construct an object containing send tag-specific data.
+   *
+   * @param[out] buffer   a raw pointer to the received data.
+   * @param[in]  length   the size in bytes of the tag message to be received.
+   * @param[in]  tag      the tag to match.
+   * @param[in]  tagMask  the tag mask to use (only used for receive operations).
+   */
+  explicit TagReceive(decltype(_buffer) buffer,
+                      const decltype(_length) length,
+                      const decltype(_tag) tag,
+                      const decltype(_tagMask) tagMask);
+
+  TagReceive() = delete;
+};
+
+class TagMultiSend {
+ public:
+  const std::vector<void*> _buffer{};   ///< Raw pointers where data to be sent is stored.
+  const std::vector<size_t> _length{};  ///< Lengths of messages.
+  const std::vector<int> _isCUDA{};     ///< Flags indicating whether the buffer is CUDA or not.
+  const ::ucxx::Tag _tag{0};            ///< Tag to match
+
+  /**
+   * @brief Constructor for send multi-buffer tag-specific data.
+   *
+   * Construct an object containing tag/multi-buffer tag-specific data.
+   *
+   * @param[in] buffer  a raw pointer to the data to be sent.
+   * @param[in] length  the size in bytes of the tag message to be sent.
+   * @param[in] tag     the tag to match.
+   */
+  explicit TagMultiSend(const decltype(_buffer)& buffer,
+                        const decltype(_length)& length,
+                        const decltype(_isCUDA)& isCUDA,
+                        const decltype(_tag) tag);
+
+  TagMultiSend() = delete;
+};
+
+class TagMultiReceive {
+ public:
+  const ::ucxx::Tag _tag{0};          ///< Tag to match
+  const ::ucxx::TagMask _tagMask{0};  ///< Tag mask to use
+
+  /**
+   * @brief Constructor for receive multi-buffer tag-specific data.
+   *
+   * Construct an object containing receive multi-buffer tag-specific data.
+   *
+   * @param[in]  tag      the tag to match.
+   * @param[in]  tagMask  the tag mask to use (only used for receive operations).
+   */
+  explicit TagMultiReceive(const decltype(_tag) tag, const decltype(_tagMask) tagMask);
+
+  TagMultiReceive() = delete;
+};
+
+using RequestData = std::variant<std::monostate,
+                                 AmSend,
+                                 AmReceive,
+                                 StreamSend,
+                                 StreamReceive,
+                                 TagSend,
+                                 TagReceive,
+                                 TagMultiSend,
+                                 TagMultiReceive>;
+
+template <class... Ts>
+struct dispatch : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+dispatch(Ts...) -> dispatch<Ts...>;
+template <class... Ts>
+dispatch(Ts&...) -> dispatch<Ts...>;
```

Comment:
I admit to not know a lot of template deduction hint magic. Did you need `Ts...` and `Ts&...` to handle overloads that had a combination of "by reference" and "by value" params?